File: topological-sort/topological_sort.php
<?php
function dfs($node, $graph, &$visited, &$stack) {
    $visited[$node] = true;
    
    if (isset($graph[$node])) {
        foreach ($graph[$node] as $neighbor) {
            if (!isset($visited[$neighbor])) {
                dfs($neighbor, $graph, $visited, $stack);
            }
        }
    }
    
    array_push($stack, $node);
}

function topologicalSort($graph) {
    $visited = [];
    $stack = [];
    
    foreach (array_keys($graph) as $node) {
        if (!isset($visited[$node])) {
            dfs($node, $graph, $visited, $stack);
        }
    }
    
    return array_reverse($stack);
}

$graph = [
    0 => [1, 2],
    1 => [3],
    2 => [3],
    3 => []
];

$result = topologicalSort($graph);
print_r($result);
?>

File: topological-sort/TopologicalSort.kt
fun topologicalSort(graph: Map<Int, List<Int>>): List<Int> {
    val visited = mutableSetOf<Int>()
    val stack = mutableListOf<Int>()
    
    fun dfs(node: Int) {
        if (node in visited) return
        visited.add(node)
        
        graph[node]?.forEach { neighbor ->
            dfs(neighbor)
        }
        
        stack.add(node)
    }
    
    graph.keys.forEach { node ->
        dfs(node)
    }
    
    return stack.reversed()
}

fun main() {
    val graph = mapOf(
        0 to listOf(1, 2),
        1 to listOf(3),
        2 to listOf(3),
        3 to emptyList()
    )
    
    println(topologicalSort(graph))
}
